// @slidev/cli/dist/build-6DZCUTHS.mjs
import {
  ViteSlidevPlugin,
  getIndexHtml,
  mergeViteConfigs
} from "./chunk-M3AHQMLP.mjs";
import "./chunk-LOUKLO2C.mjs";
import "./chunk-AQQIBD5X.mjs";

// node/commands/build.ts
import { resolve } from "node:path";
import http from "node:http";
import fs from "fs-extra";
import { mergeConfig, build as viteBuild } from "vite";
import connect from "connect";
import sirv from "sirv";
async function build(options, viteConfig = {}, args) {
  const indexPath = resolve(options.userRoot, "index.html");
  let originalIndexHTML;
  if (fs.existsSync(indexPath))
    originalIndexHTML = await fs.readFile(indexPath, "utf-8");
  await fs.writeFile(indexPath, await getIndexHtml(options), "utf-8");
  let config = void 0;
  try {
    let inlineConfig = await mergeViteConfigs(
      options,
      viteConfig,
      {
        root: options.userRoot,
        plugins: [
          {
            name: "resolve-config",
            configResolved(_config) {
              config = _config;
            }
          }
        ],
        build: {
          chunkSizeWarningLimit: 2e3
        }
      },
      "build"
    );
    inlineConfig = mergeConfig(
      inlineConfig,
      {
        plugins: [
          await ViteSlidevPlugin(options, inlineConfig.slidev || {})
        ]
      }
    );
    await viteBuild(inlineConfig);
  } finally {
    if (originalIndexHTML != null)
      await fs.writeFile(indexPath, originalIndexHTML, "utf-8");
    else
      await fs.unlink(indexPath);
  }
  const outDir = resolve(options.userRoot, config.build.outDir);
  await fs.copyFile(resolve(outDir, "index.html"), resolve(outDir, "404.html"));
  const redirectsPath = resolve(outDir, "_redirects");
  if (!fs.existsSync(redirectsPath))
    await fs.writeFile(redirectsPath, `${config.base}*    ${config.base}index.html   200
`, "utf-8");
  if ([true, "true", "auto"].includes(options.data.config.download)) {
    const { exportSlides, getExportOptions } = await import("./export-7BCUWALW.mjs");
    const port = 12445;
    const app = connect();
    const server = http.createServer(app);
    app.use(
      config.base,
      sirv(outDir, {
        etag: true,
        single: true,
        dev: true
      })
    );
    server.listen(port);
    await exportSlides({
      port,
      base: config.base,
      ...getExportOptions(args, options, outDir, "slidev-exported.pdf")
    });
    server.close();
  }
}
export {
  build
};
